Bluetooth: Classic: AVRCP: Support vendor dependent cmd

makeshi opened the pull request:
This patch introduces support for AVRCP vendor-dependent commands and responses, including full handling of fragmented messages.

makeshi committed:
bluetooth: avrcp: Add vendor-dependent with fragmentation support

This patch introduces support for AVRCP vendor-dependent commands and
responses, including full handling of fragmented messages.

- Adds fragmentation and reassembly logic for AVRCP vendor-dependent
- Introduces TX queue management using delayed work for TG
- Adds support for GetCapabilities PDUs
- Add new Kconfig for vendor-dependent with fragmentation support

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi committed:
Bluetooth: AVRCP: Add support for notification events

This patch adds AVRCP notification event handling for both CT and
TG roles. Also add Shell command support for testing notification
registration and responses.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi committed:
Bluetooth: AVRCP: Add Player app Settings vendor-dependent commands

Implement support for AVRCP Player app Settings PDUs on both CT/TG
- Added APIs for sending and handling commands like LIST_ATTRS,
   LIST_VALS, GET_CURR_VAL, SET_VAL, and related text/inform PDUs.
- Added callbacks for TG request handling and CT response processing.
- Add shell commands for Player app Settings commands testing.

Signed-off-by: Make Shi <make.shi@nxp.com>

MarkWangChinese reviewed (commented):
Only reviewed part of the codes.

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -257,6 +257,59 @@ typedef enum __packed {
+	 *
+	 * @note This is an AV/C Response semantics (not an AVRCP Error Status).
+	 *       When the CT-side callback reports REJECTED, the application MUST parse
+	 *       the net buffer to extract the 1-octet Error Status from the PDU payload.
```
Can we follow the spec definitions?
BTW, the range of these responses should be in 0x8-0xf
<img width="701" height="699" alt="image" src="https://github.com/user-attachments/assets/885bde0e-4045-4dbb-b0f2-948600e7d57b" />

makeshi replied:
I'm trying to extend the status field. I want to append the response codes (ranging from 0x8 to 0xF) to the end of the status, starting from 0xFF bt_avrcp_rsp_t.
 The response definitions will still remain as they are. pls see bt_avrcp_rsp_t
Now I'm facing a situation where a vendor-dependent Transaction Group (TG) reply requires both a response code and a status. For example, when the TG replies with an AV/C REJECTED response, it also needs to include a reason in the status.
At the same time, I feel that keeping both the response code and the status for the application might introduce redundant information, making the function somewhat awkward.
take example, if not   extend the status field, we need
 int bt_avrcp_tg_send_add_to_now_playing_rsp(struct bt_avrcp_tg *tg, uint8_t tid, uint8_t rsp_code,  uint8_t status);
If we the status field, application only need take care the status, then stack will parse it to rsp_code and status.

detail see spec:
6.15.3 Status and error codes
The responses of AVRCP Specific Browsing Commands contain a status field to indicate success/failure 
and an error status code is added to the AV/C REJECTED response if the TG rejected an AVRCP Specific AV/C command. 
It is useful for CT to know why the command was rejected by the TG. Table 6.49 shows the status and error codes to
 be used with both AVRCP Specific Browsing Commands and AVRCP Specific AV/C Commands.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -41,17 +41,64 @@ struct bt_avrcp {
+	struct k_sem lock;
+
+	/* TX work */
+	struct k_work_delayable tx_work;
```
done

makeshi commented on subsys/bluetooth/host/classic/avrcp_internal.h:
```diff
@@ -132,6 +100,7 @@ struct bt_avrcp_header {
 } __packed;
 
 struct bt_avrcp_avc_pdu {
```
done

makeshi commented on subsys/bluetooth/host/classic/avrcp_internal.h:
```diff
@@ -58,66 +58,27 @@ typedef enum __packed {
+	uint16_t received_len;			/**< Length already received */
+};
+
+struct bt_avrcp_tg_tx {
```
done

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1628,751 +1927,1493 @@ static void avrcp_register_notification_cmd_handler(struct bt_avrcp *avrcp, uint
+	return err;
+}
+
+int bt_avrcp_ct_get_cap(struct bt_avrcp_ct *ct, uint8_t tid, uint8_t cap_id)
```
Same comments for other commands.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1628,751 +1927,1493 @@ static void avrcp_register_notification_cmd_handler(struct bt_avrcp *avrcp, uint
-	    (subunit_id != BT_AVRCP_SUBUNIT_ID_ZERO)) {
-		LOG_ERR("Invalid vendor dependent command");
-		error_code = BT_AVRCP_STATUS_INVALID_PARAMETER;
+	if ((avrcp_tg_cb == NULL) || (avrcp_tg_cb->set_absolute_volume_req == NULL)) {
+		error_code = BT_AVRCP_STATUS_INTERNAL_ERROR;
```
Same comments for other commands.

makeshi committed:
Bluetooth: AVCTP: Remove redundant length check in avctp_tx_processor

Remove unnecessary checking for net_buf_tailroom(buf) < chunk_size
before sending data in avctp_tx_processor.

Signed-off-by: Make Shi <make.shi@nxp.com>

MarkWangChinese commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -505,6 +505,17 @@ config BT_AVRCP_CONTROLLER
+config BT_AVRCP_VENDOR_RX_DATA_BUF_SIZE
+	int "AVRCP Vendor Dependent RX data buffer Size"
+	default 1024
+	range 1024 $(INT16_MAX)
```
Suggest to change as `512 $(INT16_MAX)` because the `requestcontinue` is used when data length is bigger than `512`.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1093,51 +1131,34 @@ static int bt_avrcp_tg_send_vendor_rsp(struct bt_avrcp_tg *tg, uint8_t tid, uint
 	}
 
-	avrcp_ct_cb->get_caps_rsp(get_avrcp_ct(avrcp), tid, rsp);
+	avrcp_ct_cb->get_caps_rsp(get_avrcp_ct(avrcp), tid, status, buf);
```
If it is a malformed packet, I think it should not be notified.

lylezhu2012 commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -505,6 +505,17 @@ config BT_AVRCP_CONTROLLER
 	help
 	  This option enables the AVRCP profile controller function
 
+config BT_AVRCP_VENDOR_RX_DATA_BUF_SIZE
```
```suggestion
config BT_AVRCP_VD_RX_SIZE
```

or

```suggestion
config BT_AVRCP_VENDOR_DEPENDENT_RX_SIZE
```

makeshi replied:
Done by using BT_AVRCP_VD_RX_SIZE

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1131,8 +1133,151 @@ static void process_get_caps_rsp(struct bt_avrcp *avrcp, uint8_t tid, bt_avrcp_r
 }
 
+static void process_register_notification_rsp(struct bt_avrcp *avrcp, uint8_t tid,
+					      uint8_t ctype_or_rsp, struct net_buf *buf)
```
fixed, will call  notification_rsp() callback for interim and error statuses, or return NOT_IMPLEMENTED.

makeshi commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -762,6 +827,150 @@ static int cmd_get_cap(const struct shell *sh, int32_t argc, char *argv[])
+			data.identifier = 111111;
+		}
+		goto done;
+	}
```
done

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1131,8 +1133,151 @@ static void process_get_caps_rsp(struct bt_avrcp *avrcp, uint8_t tid, bt_avrcp_r
+		/* Mark as registered on interim response */
+		ct->ct_notify[event_id].registered = 1;
+		return;
+	}
```
call notification_rsp() callback for interim and error statuses, or return NOT_IMPLEMENTED.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -349,6 +407,44 @@ static inline struct bt_avrcp_tg *get_avrcp_tg(struct bt_avrcp *avrcp)
+		break;
+	}
+
+	*out_status = status;
```
I think the `status` can be removed and pass value to `*out_status` directly in switch...case

MarkWangChinese commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -2273,6 +2482,60 @@ int bt_avrcp_ct_set_browsed_player(struct bt_avrcp_ct *ct, uint8_t tid, uint16_t
+	}
+
+	memset(&ct->ct_notify[event_id], 0, sizeof(ct->ct_notify[0]));
+	ct->ct_notify[event_id].cb = cb;
```
Is it needed: clear the `.cb` when the event changed response is received. When registering one event that is already registered return error here.

makeshi replied:
Done

MarkWangChinese commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -2475,6 +2738,185 @@ int bt_avrcp_tg_send_set_browsed_player_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+		return -ENOTSUP;
+	}
+
+	switch (status) {
```
Is it needed: save the flag whether the event is registered, then return error if not registered by CT?

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -556,104 +652,686 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
+			LOG_WRN("Not enough tailroom: %u < %u", net_buf_tailroom(buf), data_len);
+			net_buf_unref(buf);
+			return -ENOMEM;
+		}
```
if the tail room of the allocated buffer is checked in function `avrcp_prepare_vendor_pdu()`, it could be omitted here.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -556,104 +652,686 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
+						 uint8_t pdu_id)
+{
+	tid++;
+	tid &= 0x0F;
```
Since the TID is managed by the application, the changing of TID will make the conflict of TID.

makeshi replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1649,3 +2504,64 @@ int bt_avrcp_tg_send_passthrough_rsp(struct bt_avrcp_tg *tg, uint8_t tid, bt_avr
+	}
+
+	rsp_code = bt_avrcp_status_to_rsp(pdu_id, status);
+	if (rsp_code == BT_AVRCP_RSP_REJECTED) {
```
How about if the `buf` is NULL when the condition is false?

makeshi replied:
It should return an error in the failure case.
If buf is NULL when the condition is true, that is a valid case, and these two scenarios have already been handled.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1649,3 +2504,64 @@ int bt_avrcp_tg_send_passthrough_rsp(struct bt_avrcp_tg *tg, uint8_t tid, bt_avr
 }
+
+static int bt_avrcp_tg_send_vendor_dependent_rsp(struct bt_avrcp_tg *tg, uint8_t tid, uint8_t pdu_id
+						 , uint8_t status, struct net_buf *buf)
```
I think the implementation of the function is hard to read. It is better to refactor it to make it more clearly.

makeshi replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1098,12 +1102,182 @@ static void process_get_caps_rsp(struct bt_avrcp *avrcp, uint8_t tid, uint8_t rs
+
+	if (rsp_code == BT_AVRCP_RSP_INTERIM) {
+		/* Mark as interim_received flag on interim response */
+		ct->ct_notify[event_id].interim_received = 1;
```
If is there a case that the `BT_AVRCP_RSP_INTERIM` is received but `BT_AVRCP_RSP_INTERIM` has been received?

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1098,12 +1102,182 @@ static void process_get_caps_rsp(struct bt_avrcp *avrcp, uint8_t tid, uint8_t rs
+	    (rsp_code == BT_AVRCP_RSP_CHANGED)) {
+		ct->ct_notify[event_id].cb(event_id, (struct bt_avrcp_event_data *)event_data);
+		ct->ct_notify[event_id].interim_received = 0;
+	}
```
```suggestion
	if ((ct->ct_notify[event_id].interim_received == 1) && (rsp_code == BT_AVRCP_RSP_CHANGED)) {
		ct->ct_notify[event_id].interim_received = 0;

		if (ct->ct_notify[event_id].cb != NULL) {
			bt_avrcp_notification_cb_t cb;
		
			cb = ct->ct_notify[event_id].cb;
			ct->ct_notify[event_id].cb = NULL;
			cb(event_id, (struct bt_avrcp_event_data *)event_data);
		}
	}
```

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -2273,6 +2482,60 @@ int bt_avrcp_ct_set_browsed_player(struct bt_avrcp_ct *ct, uint8_t tid, uint16_t
+		return -EINVAL;
+	}
+
+	memset(&ct->ct_notify[event_id], 0, sizeof(ct->ct_notify[0]));
```
I think the event of event_id is needed to be checked. If the event_id has been registered, error `-EBUSY` should be returned.

makeshi replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -2475,6 +2738,185 @@ int bt_avrcp_tg_send_set_browsed_player_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+		LOG_ERR("Failed to send notification response (err: %d)", err);
+		net_buf_unref(buf);
+	}
+	return err;
```
If the `INTERIM` has been sent and if there is any error when sending the 2th response, how to avoid the consistency state between CT and TG? Recover `tg->interim_sent[event_id]`?

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -2475,6 +2738,185 @@ int bt_avrcp_tg_send_set_browsed_player_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+		if (err < 0) {
+			net_buf_unref(buf);
+			return err;
+		}
```
Is it possible the buf is provided by the application? And AVRCP provides a series of auxiliary functions for packing the event data.

lylezhu2012 reviewed (commented):
The comments only for first 3 commits.

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -330,8 +330,8 @@ static void avctp_tx_processor(struct k_work *item)
-	if (net_buf_tailroom(buf) < chunk_size) {
-		LOG_WRN("Not enough tailroom for AVCTP payload (len: %d)", chunk_size);
+	if (buf->len < chunk_size) {
+		LOG_WRN("AVCTP payload too short: have %d, need %d", buf->len, chunk_size);
```
Combine with the comments for Lyle below:
If user_data->sent_len is 0 and buf->len < chunk_size, it should be a single packet. This condition is not checked.
Or, if user_data->sent_len != 0U, then buf->len < chunk_size should always be false since
chunk_size = MIN(buf->len, max_payload_size - BT_AVCTP_HDR_SIZE_CONTINUE_END);
The check here is redundant always in before and after before the send_fragmented_avctp.

makeshi commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -19,6 +19,7 @@
 #include <zephyr/bluetooth/hci.h>
 #include <zephyr/bluetooth/bluetooth.h>
 #include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/classic/avrcp.h>
```
This changed reason is because l2cap_br.c include avrcp_internal.h, now have removed it.
Now add the plublic avrcp.h include into avrcp_internal.h to avoid build problem, like avdtp_internal.h and rfcomm_internal.h.
And it should there is no risk of duplicate inclusion because avrcp.h have  Include Guards Protection Mechanism.

makeshi commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -505,6 +505,17 @@ config BT_AVRCP_CONTROLLER
+config BT_AVRCP_VENDOR_RX_DATA_BUF_SIZE
+	int "AVRCP Vendor Dependent RX data buffer Size"
+	default 1024
+	range 1024 $(INT16_MAX)
```
It is should 512 as Mark comments, because  this is based on AVRCP’s “Request Continue” mechanism, which triggers when the PDU exceeds 512 bytes. So, if your buffer is smaller than 512, then have risk unnecessary fragmentation handling.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -556,104 +652,686 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
+		}
+	}
+	__ASSERT(false, "Unknown PDU ID: 0x%02X", pdu_id);
+	return (bt_avrcp_ctype_t)-1;
```
Done

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -556,104 +652,686 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
+					  get_avrcp_ct(avrcp)->reassembly_buf);
+
+		/* Clean up fragmentation context */
+		cleanup_fragmentation_context(get_avrcp_ct(avrcp));
```
Spec do not support A TG shall not interleave fragmented PDUs. Once a TG has sent a start fragment it shall only send further fragments of that PDU until that PDU is completed or aborted. If a CT receives a start fragment or non-fragmented AVRCP Specific AV/C message when it already has an incomplete fragment from that TG then the CT shall consider the first PDU aborted.
bt_avrcp_ct_send_req_continuing_rsp(get_avrcp_ct(avrcp), tid, pdu->pdu_id);
If the TG follows the AVRCP specification correctly, there should never be a case where the PDU ID of the start packet is not aligned with the PDU ID of the continue or end packet.
However, if the remote side deliberately creates such a situation and ignores the AVRCP protocol, the CT could encounter this case.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1649,3 +2504,64 @@ int bt_avrcp_tg_send_passthrough_rsp(struct bt_avrcp_tg *tg, uint8_t tid, bt_avr
+			return -ENOBUFS;
+		}
+
+		if (AVRCP_STATUS_IS_REJECTED(status) == true) {
```
Done

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -2475,6 +2738,185 @@ int bt_avrcp_tg_send_set_browsed_player_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+		break;
+	case BT_AVRCP_STATUS_NOT_IMPLEMENTED:
+		rsp_code = BT_AVRCP_RSP_NOT_IMPLEMENTED;
+		break;
```
Fixed

makeshi committed:
Bluetooth: AVRCP: Implement full support vendor-dependent commands

Add full support for all remaining AVRCP vendor-dependent PDUs on both
Controller (CT) and Target (TG), including APIs, callbacks, and shell
commands for testing.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -797,6 +817,34 @@ struct bt_avrcp_ct_cb {
+	 *       first be registered using @ref bt_avrcp_notification_cb_t.
+	 */
+	void (*notification_rsp)(struct bt_avrcp_ct *ct, uint8_t tid, uint8_t status,
+				 uint8_t event_id, struct bt_avrcp_event_data *data);
```
Actually, doing it this change has advantages but also drawbacks. With this approach, the interim notification has two callbacks: one for status and one for data, also confusing user In fact, it’s similar to what I have in my shell — although the data comes through two callbacks, it can be unified into a single function for handling, and it’s easy to know whether it’s interim or changed.

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -300,7 +320,68 @@ struct bt_avrcp_passthrough_rsp {
+
+	/** Error Response */
+	BT_AVRCP_PDU_ID_GENERAL_REJECT = 0xa0,
+} bt_avrcp_pdu_id_t;
```
I think it is better to improve doxygen comments.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -86,6 +115,25 @@ static struct bt_avctp_server avctp_server;
+	for (int i = 0; i < CONFIG_BT_MAX_CONN; i++) {
+		struct bt_avrcp_tg *tg = &bt_avrcp_tg_pool[i];
+
+		if (tg->avrcp && tg->avrcp->acl_conn) {
```
```suggestion
		if ((tg->avrcp != NULL) && (tg->avrcp->acl_conn != NULL)) {
```

makeshi replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -563,97 +656,699 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
+			LOG_ERR("Failed to send rsp_buf");
+			net_buf_unref(rsp_buf);
+		}
+		atomic_test_and_clear_bit(tx->flags, BT_AVRCP_TG_FLAG_ABORT_CONTINUING);
```
I think the flag could be clear at line 977 when checking if the flag has been set.

makeshi replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -563,97 +656,699 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
-	}
+	case BT_AVRCP_PKT_TYPE_CONTINUE:
+		/* Continuation fragment */
+		if ((NULL == get_avrcp_ct(avrcp)->reassembly_buf) ||
```
```suggestion
		if ((get_avrcp_ct(avrcp)->reassembly_buf == NULL) ||
```

makeshi replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1502,6 +1685,37 @@ static void process_get_caps_cmd(struct bt_avrcp *avrcp, uint8_t tid, uint8_t ct
+	}
+
+	cmd = net_buf_pull_mem(buf, sizeof(*cmd));
+	if (cmd->event_id > BT_AVRCP_EVT_VOLUME_CHANGED) {
```
Also, if is the condition changed to `cmd->event_id > ARRAY_SIZE(get_avrcp_tg(avrcp)->tg_notify)`?

makeshi replied:
updated

jialu522 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -327,6 +408,308 @@ struct bt_avrcp_set_browsed_player_rsp {
+ *  @note The callback implementation should not block or perform heavy operations.
+ *       If needed, defer processing to another thread or task.
+ */
+typedef void(*bt_avrcp_notify_changed_cb_t)(uint8_t event_id, struct bt_avrcp_event_data *data);
```
Can we add a CT pointer in the callback function to identify which device the notification is from? The reason is: the local device can be connected to multiple devices, and all of them use the same callback function. When the application receives a notification, it has no way to determine which device it came from.

makeshi replied:
OK, will add it.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -563,97 +656,699 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
+				       BT_AVRCP_CTYPE_CONTROL, op_pdu_id,
+				       sizeof(target_pdu_id));
+	if (buf == NULL) {
+		return -ENOMEM;
```
Fixed

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -2492,6 +2768,201 @@ int bt_avrcp_tg_send_set_browsed_player_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+		return -ENOTSUP;
+	}
+
+	if (tg->tg_notify[event_id].registered == false) {
```
Done

jialu522 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -571,1081 +668,3736 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
+
+	if (avrcp->browsing_session.br_chan.chan.conn != NULL) {
+		/* If browsing session is still active, disconnect it first */
+		err = bt_avrcp_browsing_disconnect(conn);
```
If the CONFIG_BT_AVRCP_BROWSING is not enabled, this may cause a compilation error.

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -2094,6 +2214,149 @@ static void process_inform_batt_status_of_ct_cmd(struct bt_avrcp *avrcp, uint8_t
+					      BT_AVRCP_PDU_ID_SET_ABSOLUTE_VOLUME, error_code);
+	if (err < 0) {
+		LOG_ERR("Failed to send Set Absolute Volume error response");
+	}
```
fixed

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -571,89 +664,684 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
+	tx->rsp = result;
+	tx->pdu_id = pdu_id;
+	tx->sent_len = 0U;
+	atomic_clear_bit(tx->flags, BT_AVRCP_TG_FLAG_TX_ONGOING);
```
If do all flags need to be cleared?

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -1163,6 +1526,109 @@ int bt_avrcp_tg_send_set_browsed_player_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+ *  @return 0 on success or error code.
+ */
+int bt_avrcp_tg_send_list_player_app_setting_attrs_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+						       uint8_t status, struct net_buf *buf);
```
And, for the function name, it is too long. I think it also can be shorted. Just likes as, `bt_avrcp_tg_list_player_app_setting_attrs`.

lylezhu2012 replied:
I notice there is no such duplicated information in CT functions. Such as, the function `bt_avrcp_ct_inform_displayable_char_set()` of `ct`.

makeshi replied:
Done

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -1090,6 +1094,195 @@ static int process_get_caps_rsp(struct bt_avrcp *avrcp, uint8_t tid, uint8_t rsp
+	uint8_t failed_evt = 0;
+	bool found = false;
+
+	for (uint8_t i = 0; i <= BT_AVRCP_EVT_VOLUME_CHANGED; i++) {
```
updated

makeshi commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -1751,6 +2125,388 @@ static int cmd_ct_inform_batt(const struct shell *sh, size_t argc, char **argv)
+	/* Determine which attribute set to use */
+	if (use_large_attrs) {
+		rsp->num_attrs = ARRAY_SIZE(large_media_attrs);
+		for (int i = 0; i < rsp->num_attrs; i++) {
```
updated

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -571,1081 +668,3630 @@ static int bt_avrcp_send_subunit_info(struct bt_avrcp *avrcp, uint8_t tid, uint8
-	if (!IS_TG_ROLE_SUPPORTED()) {
-		LOG_ERR("Target role not supported");
-		return -ENOTSUP;
+	if (status <= BT_AVRCP_STATUS_ADDRESSED_PLAYER_CHANGED) {
```
For our design , the BT_AVRCP_STATUS_SUCCESS equals to STATUS_OPERATION_COMPLETED, and It actually means the same thing， if user select BT_AVRCP_STATUS_SUCCESS, also means the stutus BT_AVRCP_STATUS_OPERATION_COMPLETED.
And BTW, for these threes only stutus tg apis: 
int bt_avrcp_tg_set_addressed_player(struct bt_avrcp_tg *tg, uint8_t tid, uint8_t status);
int bt_avrcp_tg_play_item(struct bt_avrcp_tg *tg, uint8_t tid, uint8_t status);
int bt_avrcp_tg_add_to_now_playing(struct bt_avrcp_tg *tg, uint8_t tid, uint8_t status);

BT_AVRCP_STATUS_SUCCESS or BT_AVRCP_STATUS_OPERATION_COMPLETED also be included in the vendor payload, just like other less than BT_AVRCP_STATUS_ADDRESSED_PLAYER_CHANGED error statuses.

For other APIs, if the status is BT_AVRCP_STATUS_SUCCESS or STATUS_OPERATION_COMPLETED, the vendor payload  will follow the response format defined in the specification. For other error statuses, the vendor payload will only  contain the status error.
